packages/swapper/src/swappers/RelaySwapper/utils/getTrade.ts

```diff
@@ -0,0 +1,365 @@
+
+        // For cross-chain: always add back app fees
+        // For same-chain: only add back if input is native currency
+        if (isCrossChain || isNativeCurrencyInput) {
+          return quote.fees.app.amount
+        }
+      }
+
+      // cross-chain or same-chain with native currency as input are not applicable
+      return '0'
+    })()
+
+    const relayerFeesBuyAssetBaseUnit = (() => {
+      const relayerFeeRelayToken = quote.fees?.relayer?.currency
+      if (relayerFeeRelayToken && isValidRelayToken(relayerFeeRelayToken)) {
+        const relayerFeesAsset = relayTokenToAsset(relayerFeeRelayToken, deps.assetsById)
+        const relayerFeeAmount = quote.fees?.relayer?.amount ?? '0'
+
+        // If fee is already in buy asset, return as is
+        if (relayerFeesAsset?.assetId === buyAsset.assetId) {
+          return relayerFeeAmount
+        }
+
+        // if fee is in sell asset, convert to buy asset
+        if (relayerFeesAsset?.assetId === sellAsset.assetId) {
+          return convertPrecision({
+            value: relayerFeeAmount,
+            inputExponent: relayerFeesAsset.precision,
+            outputExponent: buyAsset.precision,
+          })
+            .times(rate)
+            .toFixed(0)
+        }
+
+        // If fee is in a different asset, convert to buy asset
+        const feeAmountUsd = quote.fees?.relayer?.amountUsd ?? '0'
+        const buyAssetUsd = quote.details?.currencyOut?.amountUsd ?? '0'
+        const buyAssetAmountBaseUnit = quote.details?.currencyOut?.amount ?? '1'
+
+        if (feeAmountUsd && buyAssetUsd && buyAssetAmountBaseUnit) {
+          // Calculate the rate: (buyAssetAmount / buyAssetUsd) gives us "buy asset per USD"
+          // Then multiply by feeAmountUsd to get the equivalent buy asset amount
+          const buyAssetCryptoBaseUnitPerUsd = bnOrZero(buyAssetAmountBaseUnit).div(buyAssetUsd)
+          const buyAssetFeesCryptoBaseUnit = bnOrZero(feeAmountUsd).times(
+            buyAssetCryptoBaseUnitPerUsd,
+          )
+
+          return buyAssetFeesCryptoBaseUnit.toFixed(0)
+        }
+
+        return '0'
+      }
+
+      return '0'
+    })()
+
+    // Add back relayer service and gas fees (relayer is including both) since they are downsides
+    // And add appFees
+    const buyAmountBeforeFeesCryptoBaseUnit = bnOrZero(currencyOut.minimumAmount)
+      // @blocking: relayer can be ETH or destination token
+      .plus(relayerFeesBuyAssetBaseUnit)
+      .plus(appFeesBaseUnit)
+      .toFixed()
+
+    return {
+      allowanceContract: hasApprovalStep ? swapSteps[0]?.items?.[0]?.data?.to : undefined,
+      rate,
+      buyAmountBeforeFeesCryptoBaseUnit,
+      buyAmountAfterFeesCryptoBaseUnit: currencyOut.minimumAmount ?? '0',
+      sellAmountIncludingProtocolFeesCryptoBaseUnit,
+      buyAsset,
+      sellAsset,
+      accountNumber,
+      feeData: {
+        networkFeeCryptoBaseUnit: quote.fees?.gas?.amount ?? '0',
```

Comment:
Shouldn't we guard against that?

Reply:
The optional dance has been removed in [29fc44d](https://github.com/shapeshift/web/pull/9188/commits/29fc44d8ce158086f2cbec2c3c7ee0dd44302dac)

And it's now guarded in getEvmTransactionFees

Reply:
Still incorrect, but closing this comment not to keep a duplicate discussions on a stale diff.